### PR TITLE
Add aliased tables cache in query_builder.

### PR DESCRIPTION
**Pull request by tianhe1986**
Set `$_protect_identifiers` to `TRUE` and also set `dbprefix` to `ok_`.

Then I used the following code:
```
$this->db->start_cache();
$this->db->select('us.name');
$this->db->from('user as us');
$this->db->stop_cache();
$sql = $this->db->get_compiled_select('');
echo $sql."\n";
$sql = $this->db->get_compiled_select('');
echo $sql."\n";
```

and got

```
SELECT `us`.`name` FROM `ok_user` as `us`
SELECT `ok_us`.`name` FROM `ok_user` as `us`
```

The table prefix was not properly processed in `CI_DB_query_builder::_merge_cache()`. 

And the similar situation when using qb cache with alias in JOIN statement.  For example,
```
$this->db->start_cache();
$this->db->select('user.name, jo.description');
$this->db->from('user');
$this->db->join('job as jo', 'user.name = jo.name');
$this->db->stop_cache();
$sql = $this->db->get_compiled_select('');
echo $sql."\n";
$sql = $this->db->get_compiled_select('');
echo $sql."\n";
```
And got
```
SELECT `ok_user`.`name`, `jo`.`description` FROM `ok_user` JOIN `ok_job` as `jo` ON `ok_user`.`name` = `jo`.`name`
SELECT `ok_user`.`name`, `ok_jo`.`description` FROM `ok_user` JOIN `ok_job` as `jo` ON `ok_user`.`name` = `jo`.`name`
```
Because $qb_aliased_tables was reset to empty array and not be reprocessed when merging JOIN cache.

So I add aliased tables cache in this PR to fix them.

### CODE REVIEW

File: system/database/DB_query_builder.php [hunk 1]
```diff
@@ -215,6 +215,13 @@
 	protected $qb_cache_join			= array();
 
 	/**
+	 * QB Cache aliased tables list data
```
Comment (narfbg) [inline system/database/DB_query_builder.php]: Wouldn't care for this that much, but there are other changes to be made as well, so ... Keep either 'list' or 'data', one is superfluous.

File: system/database/DB_query_builder.php [hunk 2]
```diff
@@ -215,6 +215,13 @@
...
+	 *
+	 * @var	array
+	 */
+	protected $qb_cache_aliased_tables			= array();
```
Comment (narfbg) [inline system/database/DB_query_builder.php]: Just `$qb_cache_aliases`.

File: system/database/DB_query_builder.php [hunk 3]
```diff
@@ -2284,6 +2291,11 @@ protected function _track_aliases($table)
 			if ( ! in_array($table, $this->qb_aliased_tables))
 			{
 				$this->qb_aliased_tables[] = $table;
+				if ($this->qb_caching === TRUE && ! in_array($table, $this->qb_cache_aliased_tables))
```
Comment (narfbg) [inline system/database/DB_query_builder.php]: Always set `in_array()`'s third parameter to `TRUE`.

File: system/database/DB_query_builder.php [hunk 4]
```diff
@@ -2616,16 +2628,17 @@ public function stop_cache()
...
-			'qb_cache_from'			=> array(),
-			'qb_cache_join'			=> array(),
-			'qb_cache_where'		=> array(),
-			'qb_cache_groupby'		=> array(),
-			'qb_cache_having'		=> array(),
-			'qb_cache_orderby'		=> array(),
-			'qb_cache_set'			=> array(),
-			'qb_cache_exists'		=> array(),
-			'qb_cache_no_escape'	=> array()
+			'qb_cache_select'			=> array(),
```
Comment (narfbg) [inline system/database/DB_query_builder.php]: Don't re-align.

File: system/database/DB_query_builder.php [hunk 5]
```diff
@@ -2676,13 +2689,6 @@ protected function _merge_cache()
 				$this->qb_no_escape = $qb_no_escape;
 			}
 		}
-
-		// If we are "protecting identifiers" we need to examine the "from"
```
Comment (narfbg) [inline system/database/DB_query_builder.php]: Why is this removed?

File: system/database/DB_query_builder.php (hunk 2 above)
Comment (tianhe1986) [inline system/database/DB_query_builder.php]: En,  this is because QB aliased tables list named `$qb_aliased_tables`,  and in `_merge_cache()` ,
```
$qb_variable	= 'qb_'.$val;
$qb_cache_var	= 'qb_cache_'.$val;
```

File: system/database/DB_query_builder.php (hunk 5 above)
Comment (tianhe1986) [inline system/database/DB_query_builder.php]: There is no need to re-examine when using alias cache. They have been put into `$qb_aliased_tables`.

File: system/database/DB_query_builder.php (hunk 2 above)
Comment (narfbg) [inline system/database/DB_query_builder.php]: I see ... ok.
